src/stella_vslam/solve/essential_solver.cc

```diff
@@ -48,28 +49,41 @@ void essential_solver::find_via_ransac(const unsigned int max_num_iter, const bo
             min_set_bearings_2.at(i) = bearings_2_.at(matches_12_.at(idx).second);
         }
 
-        // 2-2. Compute an essential matrix
-        E_21_in_sac = compute_E_21(min_set_bearings_1, min_set_bearings_2);
-
-        // 2-3. Check inliers and compute a cost
-        float cost_in_sac;
-        unsigned int num_inliers = check_inliers(E_21_in_sac, is_inlier_match_in_sac, cost_in_sac);
+        // 2-2. Compute candidate essential matrices with the minimal solver
+        std::vector<Mat33_t> E_mats;
+        if (min_set_size == 5) {
+            E_mats = compute_E_21_minimal(min_set_bearings_1, min_set_bearings_2);
+        }
+        else {
+            E_mats.push_back(compute_E_21_nonminimal(min_set_bearings_1, min_set_bearings_2));
+        }
 
-        // 2-4. Update the best model
-        if (num_inliers > min_set_size && best_cost_ > cost_in_sac) {
-            best_cost_ = cost_in_sac;
-            best_E_21_ = E_21_in_sac;
-            is_inlier_match_ = is_inlier_match_in_sac;
+        // see if any of the candidates are better than best_E_21_
+        for (const auto& E_in_sac : E_mats) {
+            // 2-3. Check inliers and compute a cost
+            float cost_in_sac;
+            unsigned int num_inliers = check_inliers(E_in_sac, is_inlier_match_in_sac, cost_in_sac);
+
+            // 2-4. Update the best model
+            if (num_inliers > min_set_size && best_cost_ > cost_in_sac) {
+                best_cost_ = cost_in_sac;
+                best_E_21_ = E_in_sac;
+                is_inlier_match_ = is_inlier_match_in_sac;
+                best_num_inliers = num_inliers;
+            }
         }
     }
 
     solution_is_valid_ = best_cost_ < std::numeric_limits<float>::max();
 
-    if (!recompute || !solution_is_valid_) {
+    // we need a valid solution with at least 8 inliers to do the refinement
+    // since it uses the 8pt algorithm
+    if (!recompute || !solution_is_valid_ || best_num_inliers < 8) {
```

Comment:
I think solution_is_valid_ should be updated, what do you think?

```c++
    // we need a valid solution with at least 8 inliers to do the refinement
    // since it uses the 8pt algorithm
    solution_is_valid_ = best_cost_ < std::numeric_limits<float>::max() && best_num_inliers < 8;

    if (!recompute || !solution_is_valid_) {
```

Reply:
since the code can use the 5 point algorithm to find E, it is possible that the solution could be valid if we were only given 5 correct correspondences. but with this logic, it would mark such a solution as invalid (i.e. it would set solution_is_valid_ to false). 

if you wanted to be conservative and say that even though 5 correspondences is technically enough, we will not accept a solution from a set of less than 8 correct correspondences, I think this would be reasonable and the code you shared above would be a good way to do that. But I think this would justify a comment to be explicit about why that was done.

However, what I had initially intended with that check was more like this:
```
    solution_is_valid_ = best_cost_ < std::numeric_limits<float>::max();

    // even if the user wants to refine the solution, we need at least 8 inliers to do the refinement
    // since it uses the 8pt algorithm
    const bool should_recompute = (recompute && best_num_inliers >= 8);

    if (!should_recompute || !solution_is_valid_) {
```

Do you have a preference?

Reply:
Understood. I think no need to be changed.